tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc

```diff
@@ -1946,6 +1946,96 @@ static LogicalResult Verify(DataFormatVecPermuteOp op) {
+/// elements of the divisor are non-zero.
+
+/// Similarly, if the op is tf.MulNoNan and the multiplier is a constant tensor
+/// (with all the elements of any allowed type: float or complex), rewrite the
+/// op to the multiplier if all the elements of the multiplier are zero and to
+/// tf.Mul if all the elements of the multiplier are non-zero.
+
+/// Replace the given op with an op of type `RetT`. Upon calling
+/// DivNoNanOrMulNoNanConstantY for canonicalizing tf.DivNoNan, TF::DivOp is
+/// passed as the second argument and for canonicalizing tf.MulNoNan, TF::MulOp
+/// is passed as the second argument.
+template <typename OpT, typename RetT>
+class DivNoNanOrMulNoNanConstantY : public OpRewritePattern<OpT> {
+  using OpRewritePattern<OpT>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OpT op,
+                                PatternRewriter& rewriter) const override {
+    static_assert(
+        llvm::is_one_of<OpT, DivNoNanOp, MulNoNanOp>::value,
+        "only canonicalization of tf.DivNoNan and tf.MulNoNan is supported");
+
+    // A lambda function named `numZerosInAttr` is used for returning the
+    // total number of zero elements (float/complex type) in `attr`.
+    auto numZerosInAttr = [](ElementsAttr attr, bool complexElements) {
+      unsigned numZeros = 0;
+      if (!complexElements) {
+        for (const auto val : attr.getValues<APFloat>()) {
+          if (val.isZero()) numZeros++;
+        }
+      } else {
+        for (const auto val : attr.getValues<std::complex<APFloat>>()) {
+          // Note that when `val` is of complex type, it is zero iff both
+          // its real and imaginary parts are zero.
+          if (val.real().isZero() && val.imag().isZero()) numZeros++;
+        }
+      }
+      return numZeros;
+    };
+
+    // Note that `y` is the divisor if the op is tf.DivNoNan and it is the
+    // multiplier if the op is tf.MulNoNan.
+    Value y = op.y();
+    // The below if condition is true iff `y.getDefiningOp()` is of the type
+    // TF::ConstOp, i.e., if `y` is defined by an op and it is the tf.Const op.
+    // In that case, `yDefOp` stores this tf.Const op.
+    // Note that if `y` is a block argument, `y.getDefiningOp()` will return
+    // null, which will get propogated by dyn_cast_or_null to `yDefOp`.
+    // Further, if `y` is defined by an op other than tf.Const,
+    // `y.getDefiningOp()` will not return null but dyn_cast_or_null will.
+    if (auto yDefOp = dyn_cast_or_null<TF::ConstOp>(y.getDefiningOp())) {
+      Type typeOfElementsInY = getElementTypeOrSelf(y.getType());
+      ElementsAttr attr = yDefOp.value();
+      unsigned totalNumberOfElementsInY = attr.getNumElements();
+      unsigned totalNumberOfZeroElementsInY =
+          numZerosInAttr(attr, typeOfElementsInY.isa<ComplexType>());
+      if (totalNumberOfElementsInY == totalNumberOfZeroElementsInY) {
+        // When all the elements in `y` are zero, replace the op (tf.divNoNan
+        // or tf.MulNoNan) with `y`.
```

Comment:
I'm puzzled: when the divisor is all 0.0, shouldn't the result be infinity instead?

Reply:
> I'm puzzled: when the divisor is all 0.0, shouldn't the result be infinity instead?

Actually, that's not correct. Basically, the `tf.DivNoNan` op results in 0 if the divisor is 0, as can be seen described [here](https://www.tensorflow.org/mlir/tf_ops#tfdivnonan_tfdivnonanop).

Reply:
That's... interesting! I guess the "NoNan" in the name is a shortcut for "NoNanNoInf". Thanks!

Reply:
> That's... interesting! I guess the "NoNan" in the name is a shortcut for "NoNanNoInf". Thanks!

Yes, I guess! :)